src/java.base/share/classes/jdk/internal/net/quic/QuicTLSContext.java

```diff
@@ -0,0 +1,152 @@
+        boolean parametersSupported = isQuicCompatible(sslContext.getSupportedSSLParameters());
+        if (!parametersSupported) {
+            return false;
+        }
+        // horrible hack - what we do here is try and get hold of a SSLContext
+        // that has already been initialised and configured with the HttpClient.
+        // We see if that SSLContext is created using an implementation of
+        // sun.security.ssl.SSLContextImpl. Since there's no API
+        // available to get hold of that underlying implementation, we use
+        // MethodHandle lookup to get access to the field which holds that
+        // detail.
+        final Object underlyingImpl = CONTEXT_SPI.get(sslContext);
+        if (!(underlyingImpl instanceof SSLContextImpl ssci)) {
+            return false;
+        }
+        return ssci.isUsableWithQuic();
+    }
+
+    /**
+     * {@return {@code true} if protocols of the given {@code parameters} support QUIC TLS, {@code false} otherwise}
+     */
+    public static boolean isQuicCompatible(SSLParameters parameters) {
+        String[] protocols = parameters.getProtocols();
+        return protocols != null && Arrays.asList(protocols).contains("TLSv1.3");
+    }
+
+    private static SSLContextImpl getSSLContextImpl(
+            final SSLContext sslContext) {
+        final Object underlyingImpl = CONTEXT_SPI.get(sslContext);
+        assert underlyingImpl instanceof SSLContextImpl;
+        return (SSLContextImpl) underlyingImpl;
+    }
+
+    /**
+     * Constructs a QuicTLSContext for the given {@code sslContext}
+     *
+     * @param sslContext The SSLContext
+     * @throws IllegalArgumentException If the passed {@code sslContext} isn't
+     *                                  supported by the QuicTLSContext
+     * @see #isQuicCompatible(SSLContext)
+     */
+    public QuicTLSContext(final SSLContext sslContext) {
+        Objects.requireNonNull(sslContext);
+        if (!isQuicCompatible(sslContext)) {
+            throw new IllegalArgumentException(
+                    "Cannot construct a QUIC TLS context with the given SSLContext");
+        }
+        this.sslCtxImpl = getSSLContextImpl(sslContext);
+    }
+
+    /**
+     * Creates a {@link QuicTLSEngine} using this context
+     * <p>
+     * This method does not provide hints for session caching.
+     *
+     * @return the newly created QuicTLSEngine
+     */
+    public QuicTLSEngine createEngine() {
+        return createEngine(null, -1);
+    }
+
+    /**
+     * Creates a {@link QuicTLSEngine} using this context using
+     * advisory peer information.
+     * <p>
+     * The provided parameters will be used as hints for session caching.
+     * The {@code peerHost} parameter will be used in the server_name extension,
+     * unless overridden later.
+     *
+     * @param peerHost The peer hostname or IP address. Can be null.
+     * @param peerPort The peer port, can be -1 if the port is unknown
```

Comment:
Would that be the hostname in the URI, or in the AltService?
Maybe we could add an `@apiNote` here to clarify it.

Reply:
Well the javadoc here was written to match the one on SSLContext#createSSLEngine. The peer information is used for caching, but it's also used in the SNI extension, so ideally users should use the URI address, not the alt service one.

Reply:
OK. Maybe that would deserve a note - since with HTTP/3 we have potentially two addresses and two ports.